### PR TITLE
[SYCL] Add new spellings for SYCL FPGA attributes

### PR DESCRIPTION
**Pull request by smanna12**
Currently, there is a mismatch between our namespace in SYCL (which use intel::)
and the prefixes we use for attributes in LLVM.

This patch updates the attribute spellings that currently use
intelfpga:: prefixes to intel:: in all lowercase and no change for the attribute spelling that use intel:: prefixes.

The patches make this in a three step change:

1. Enable the intel:: prefix, without disabling the older versions.
2. Provide a deprecation warning for the older versions.
3. Disable intelfpga:: versions.

The affected attributes are:

•intelfpga::num_simd_work_items
•intelfpga::max_work_group_size
•intelfpga::max_global_work_dim
•intelfpga::no_global_work_offset
•intelfpga::ivdep
•intelfpga::ii
•intelfpga::max_concurrency
•intelfpga::loop_coalesce
•intelfpga::disable_loop_pipelining
•intelfpga::max_interleaving
•intelfpga::speculated_iterations
•intelfpga::doublepump
•intelfpga::singlepump
•intelfpga::memory
•intelfpga::register
•intelfpga::bankwidth
•intelfpga::numbanks
•intelfpga::private_copies
•intelfpga::merge
•intelfpga::max_replicates
•intelfpga::simple_dual_port
•intelfpga::bank_bits
•intelfpga::force_pow2_depth

1. Additionally, the patches renames spelling for the following two attributes:

•intelfpga::memory -> intel::fpga_memory
•intelfpga::register -> intel::fpga_register

2. Modify the tests and documentation.

Signed-off-by: Soumi Manna <soumi.manna@intel.com>

### CODE REVIEW

**Comment by keryell** [comment]
I think that the SYCL committee went back in the other direction last week: https://gitlab.khronos.org/sycl/Specification/-/merge_requests/537

Lowercase looks more like C++ and less like C macros.

**Comment by Fznamznon** [comment]
> I think that the SYCL committee went back in the other direction last week: https://gitlab.khronos.org/sycl/Specification/-/merge_requests/537
> 
> Lowercase looks more like C++ and less like C macros.

@jbrodman could you please comment on this?

**Review by erichkeane — APPROVED** [review]

**Comment by erichkeane** [comment]
Code looks fine, @jbrodman / @mkinsner  need to comment on this, our feature request was given telling us to do all capitals, but if the spec changed we should use this opportunity to go back to lowercase.

**Comment by keryell** [comment]
By the way, @erichkeane since Intel is more motivated than me to have attributes everywhere (even if for any FPGA companies, it looks like any problem is solved by adding a `#pragma` or an attribute :-) ), would you be kind enough to write a paper for ISO C++23 to have multiple levels of namespace in the attributes, such as allowing things like `[[sycl::intel::num_simd_work_items]]`. Right now only 1 level is allowed.

**Comment by erichkeane** [comment]
> 
> 
> By the way, @erichkeane since Intel is more motivated than me to have attributes everywhere (even if for any FPGA companies, it looks like any problem is solved by adding a `#pragma` or an attribute :-) ), would you be kind enough to write a paper for ISO C++23 to have multiple levels of namespace in the attributes, such as allowing things like `[[sycl::intel::num_simd_work_items]]`. Right now only 1 level is allowed.

Not a bad idea.  We're about to have someone on board who can help with that, and is an attribute expert (and managed it through WG14), so I'll see what he thinks.

**Comment by smanna12** [comment]
> Code looks fine, @jbrodman / @mkinsner need to comment on this, our feature request was given telling us to do all capitals, but if the spec changed we should use this opportunity to go back to lowercase.

@mkinsner could you please comment on this? Should we bring the fpga specific ones under lowercase `intel::` now?

**Comment by mkinsner** [comment]
> > Code looks fine, @jbrodman / @mkinsner need to comment on this, our feature request was given telling us to do all capitals, but if the spec changed we should use this opportunity to go back to lowercase.
> 
> @mkinsner could you please comment on this? Should we bring the fpga specific ones under lowercase `intel::` now?

Probably, but some details are being settled.  Please contact me offline.

**Comment by mkinsner** [comment]
> By the way, @erichkeane since Intel is more motivated than me to have attributes everywhere

A number of these extensions should migrate to non-attribute representations, with attributes like these used to implement the abstraction behind the scenes when appropriate.  I think that these should be a stepping stone.  I've been pushing the non-attribute style internally within Intel FPGA recently.

**Comment by smanna12** [comment]
I have updated patch (Moving the intelfpga -> intel:: and no change for intel:: attribute spelling) and summary above based on the discussion offline with @erichkeane @mkinsner @jbrodman.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 1]
```diff
@@ -5279,6 +5329,13 @@ static void handleIntelFPGARegisterAttr(Sema &S, Decl *D,
   if (checkIntelFPGARegisterAttrCompatibility(S, D, Attr))
     return;
 
+  if (Attr.getAttributeSpellingListIndex() ==
+      IntelFPGARegisterAttr::CXX11_intelfpga_register) {
+    S.Diag(Attr.getLoc(), diag::warn_attribute_spelling_deprecated) << Attr;
+    S.Diag(Attr.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::fpga_register'";
+  }
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaDeclAttr.cpp:5337]: It there a way to move this part of code to a function? Because it is repeated many times.

File: clang/test/SemaSYCL/intel-fpga-loops.cpp [hunk 2]
```diff
@@ -3,148 +3,177 @@
...
+  [[intel::max_interleaving(4)]] int i[10];
   // expected-error@+1 {{intelfpga loop attributes must be applied to for, while, or do statements}}
-  [[intelfpga::speculated_iterations(6)]] int j[10];
+  [[intel::speculated_iterations(6)]] int j[10];
+}
+
+// Test for deprecated spelling of Intel FPGA loop attributes
+void foo_deprecated() {
+  int a[10];
+  // expected-warning@+2 {{attribute 'ivdep' is deprecated}}
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/intel-fpga-loops.cpp:33]: I think this diagnostic is a bit strange because the attribute mention doesn't match with actual spelling in the code. Can it say that spelling is deprecated or include the full spelling string as it was written in source code, i.e. like this:
```suggestion
  // expected-warning@+2 {{attribute 'intelfpga::ivdep' is deprecated}}
```

File: clang/test/CodeGenSYCL/device-indirectly-callable-attr.cpp [hunk 3]
```diff
@@ -2,8 +2,7 @@
 
 void helper() {}
 
-[[intel::device_indirectly_callable]]
-void foo() {
+[[intel::device_indirectly_callable]] void foo() {
   helper();
 }
```
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/device-indirectly-callable-attr.cpp:7]: This change seems a bit unrelated.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 1 above)
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:5337]: @Fznamznon, I thought about to move this in a separate function. The issue is that we are telling users to use different attribute spelling in note instead of the deprecated attribute one , not sure how can we do it. Any thought?

Comment (Fznamznon) [reply clang/lib/Sema/SemaDeclAttr.cpp:5337]: I don't know how to get a string for desired spelling from the attribute object, unfortunately (maybe @erichkeane can suggest something?). If there is no way to do it, we could just pass a string with desired spelling as a function parameter.

Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp:5337]: We don't have a good way to do that unfortunately, best we  have is Attr::getSpelling, which gets the current spelling.

Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp:5337]: >@Fznamznon, I thought about to move this in a separate function. The issue is that we are telling users to use different attribute spelling in note instead of the deprecated attribute one , not sure how can we do it. Any thought?

Could either be passed in as a parameter along with the attribute kind and spelling, or we could just have one function that checks ALL of them?

File: clang/test/CodeGenSYCL/device-indirectly-callable-attr.cpp (hunk 3 above)
Comment (smanna12) [reply clang/test/CodeGenSYCL/device-indirectly-callable-attr.cpp]: Yes, this is unrelated. Test is updated.

File: clang/test/SemaSYCL/intel-fpga-loops.cpp (hunk 2 above)
Comment (smanna12) [reply clang/test/SemaSYCL/intel-fpga-loops.cpp]: Done

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 1 above)
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Thanks for the suggestion. I have moved the diagnostic in a separate function.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 4]
```diff
@@ -318,6 +318,15 @@ static bool checkAttrMutualExclusion(Sema &S, Decl *D, const Attr &AL) {
   return false;
 }
 
+static bool checkDeprecatedFPGASYCLSpelling(Sema &S, const ParsedAttr &Attr) {
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaDeclAttr.cpp:321]: How about 
```suggestion
static bool checkDeprecatedSYCLAttributeSpelling(Sema &S, const ParsedAttr &Attr) {
```

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 5]
```diff
@@ -318,6 +318,15 @@ static bool checkAttrMutualExclusion(Sema &S, Decl *D, const Attr &AL) {
...
+  if (Attr.getScopeName()->isStr("intelfpga")) {
+    S.Diag(Attr.getLoc(), diag::warn_attribute_spelling_deprecated)
+        << "'" + Attr.getNormalizedFullName() + "'";
+    return true;
+  }
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaDeclAttr.cpp:326]: We can make it lazier
```suggestion
  if (Attr.getScopeName()->isStr("intelfpga"))
    return S.Diag(Attr.getLoc(), diag::warn_attribute_spelling_deprecated)
        << "'" + Attr.getNormalizedFullName() + "'";
```

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 6]
```diff
@@ -3002,6 +3019,12 @@ static void handleNumSimdWorkItemsAttr(Sema &S, Decl *D,
   if (D->getAttr<SYCLIntelNumSimdWorkItemsAttr>())
     S.Diag(Attr.getLoc(), diag::warn_duplicate_attribute) << Attr;
 
+  if (checkDeprecatedFPGASYCLSpelling(S, Attr)) {
+    S.Diag(Attr.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::num_simd_work_items'";
+    return;
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaDeclAttr.cpp:3025]: So, if an attribute has deprecated spelling we emit a warning and don't add this attribute to a declaration, right?

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 7]
```diff
@@ -5192,6 +5227,18 @@ static void handleIntelFPGAPumpAttr(Sema &S, Decl *D, const ParsedAttr &Attr) {
     D->addAttr(IntelFPGAMemoryAttr::CreateImplicit(
         S.Context, IntelFPGAMemoryAttr::Default));
 
+  if (Attr.getKind() == ParsedAttr::AT_IntelFPGADoublePump &&
+      checkDeprecatedFPGASYCLSpelling(S, Attr)) {
+    S.Diag(Attr.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::doublepump'";
+    return;
+  } else if (Attr.getKind() == ParsedAttr::AT_IntelFPGASinglePump &&
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaDeclAttr.cpp:5235]: Please don't do else after return.
```suggestion
    return;
  }
if (Attr.getKind() == ParsedAttr::AT_IntelFPGASinglePump &&
```

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 8]
```diff
@@ -5298,6 +5357,18 @@ static void handleOneConstantPowerTwoValueAttr(Sema &S, Decl *D,
   if (checkAttrMutualExclusion<IntelFPGARegisterAttr>(S, D, Attr))
     return;
 
+  if (Attr.getKind() == ParsedAttr::AT_IntelFPGABankWidth &&
+      checkDeprecatedFPGASYCLSpelling(S, Attr)) {
+    S.Diag(Attr.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::bankwidth'";
+    return;
+  } else if (Attr.getKind() == ParsedAttr::AT_IntelFPGANumBanks &&
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaDeclAttr.cpp:5365]: And here else after return.

File: clang/lib/Sema/SemaStmtAttr.cpp [hunk 9]
```diff
@@ -74,6 +74,16 @@ static Attr *handleSuppressAttr(Sema &S, Stmt *St, const ParsedAttr &A,
       S.Context, A, DiagnosticIdentifiers.data(), DiagnosticIdentifiers.size());
 }
 
+static bool CheckForDeprecatedSYCLLoopAttributeSpelling(Sema &S,
+                                                        const ParsedAttr &A) {
+  if (A.getScopeName()->isStr("intelfpga")) {
+    S.Diag(A.getLoc(), diag::warn_attribute_spelling_deprecated)
+        << "'" + A.getNormalizedFullName() + "'";
+    return true;
+  }
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaStmtAttr.cpp:83]: Same here.
```suggestion
  if (A.getScopeName()->isStr("intelfpga"))
    return S.Diag(A.getLoc(), diag::warn_attribute_spelling_deprecated)
        << "'" + A.getNormalizedFullName() + "'";
```

File: clang/lib/Sema/SemaStmtAttr.cpp [hunk 10]
```diff
@@ -95,6 +105,36 @@ static Attr *handleIntelFPGALoopAttr(Sema &S, const ParsedAttr &A) {
...
+             CheckForDeprecatedSYCLLoopAttributeSpelling(S, A)) {
+    S.Diag(A.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::speculated_iterations'";
+    return nullptr;
+  } else if (A.getKind() == ParsedAttr::AT_SYCLIntelFPGALoopCoalesce &&
+             CheckForDeprecatedSYCLLoopAttributeSpelling(S, A)) {
+    S.Diag(A.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::loop_coalesce'";
+    return nullptr;
+  }
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaStmtAttr.cpp:136]: Elses after returns in this whole bunch of ifs.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 6 above)
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:3025]: @Fznamznon yes, correct.

Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp:3025]: We should probably add the attribute to the declaration still.  "Deprecated" typically means "This still works, but we're going to make it stop working in the future".

Comment (Fznamznon) [reply clang/lib/Sema/SemaDeclAttr.cpp:3025]: > We should probably add the attribute to the declaration still. "Deprecated" typically means "This still works, but we're going to make it stop working in the future".

Yep, that is what I'd expect from "deprecated" attribute.

Comment (MrSidims) [reply clang/lib/Sema/SemaDeclAttr.cpp:3025]: Agree that we shall have the attribute added to the decl.

Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp]: I have updated patch to fix the issue. Now this adds attribute to the declaration.  Deprecated attribute spelling works same as new spelling.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 8 above)
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Done

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 7 above)
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Done

File: clang/lib/Sema/SemaStmtAttr.cpp (hunk 10 above)
Comment (smanna12) [reply clang/lib/Sema/SemaStmtAttr.cpp]: Done

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 5 above)
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Done

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 4 above)
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Done

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 11]
```diff
@@ -3031,10 +3051,14 @@ static void handleMaxGlobalWorkDimAttr(Sema &S, Decl *D,
...
     }
   }
-
   if (D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>())
     S.Diag(Attr.getLoc(), diag::warn_duplicate_attribute) << Attr;
 
+  if (checkDeprecatedSYCLAttributeSpelling(S, Attr)) {
+    S.Diag(Attr.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::max_global_work_dim'";
+  }
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaDeclAttr.cpp:3060]: Could you please drop curly braces for one statement if conditions?
I see a lot of cases in the patch.

File: clang/lib/Sema/SemaStmtAttr.cpp [hunk 12]
```diff
@@ -95,6 +103,31 @@ static Attr *handleIntelFPGALoopAttr(Sema &S, const ParsedAttr &A) {
...
+        << "'intel::max_interleaving'";
+  } else if (A.getKind() == ParsedAttr::AT_SYCLIntelFPGASpeculatedIterations &&
+             checkDeprecatedSYCLLoopAttributeSpelling(S, A)) {
+    S.Diag(A.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::speculated_iterations'";
+  } else if (A.getKind() == ParsedAttr::AT_SYCLIntelFPGALoopCoalesce &&
+             checkDeprecatedSYCLLoopAttributeSpelling(S, A)) {
+    S.Diag(A.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::loop_coalesce'";
+  }
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaStmtAttr.cpp:129]: Is else conditions necessary here?

**Review by Fznamznon — COMMENTED** [review]
LGTM overall, except small style nits.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 11 above)
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Done

File: clang/lib/Sema/SemaStmtAttr.cpp (hunk 12 above)
Comment (smanna12) [reply clang/lib/Sema/SemaStmtAttr.cpp:129]: > Is else conditions necessary here?

I do not think that we need else here since we give separate warning/error for unknow loop attribute spellings. If you would want, I can split the "else if " to 
else {
 if (checkDeprecatedSYCLLoopAttributeSpelling(S, A))    
    S.Diag(A.getLoc(), diag::note_spelling_suggestion)
        << "'intel::loop_coalesce'";
  }

Comment (Fznamznon) [reply clang/lib/Sema/SemaStmtAttr.cpp:129]: I think I'm fine with the current state.

**Review by Fznamznon — APPROVED** [review]

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 13]
```diff
@@ -5168,6 +5190,12 @@ static void handleNoGlobalWorkOffsetAttr(Sema &S, Decl *D,
     S.Diag(Attr.getLoc(), diag::warn_boolean_attribute_argument_is_not_valid)
         << Attr;
 
+  if (Attr.getKind() == ParsedAttr::AT_SYCLIntelNoGlobalWorkOffset &&
+      checkDeprecatedSYCLAttributeSpelling(S, Attr)) {
+    S.Diag(Attr.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::no_global_work_offset'";
+  }
```
Comment (premanandrao) [inline clang/lib/Sema/SemaDeclAttr.cpp:5197]: unneeded braces

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 14]
```diff
@@ -5192,6 +5220,16 @@ static void handleIntelFPGAPumpAttr(Sema &S, Decl *D, const ParsedAttr &Attr) {
     D->addAttr(IntelFPGAMemoryAttr::CreateImplicit(
         S.Context, IntelFPGAMemoryAttr::Default));
 
+  if (Attr.getKind() == ParsedAttr::AT_IntelFPGADoublePump &&
+      checkDeprecatedSYCLAttributeSpelling(S, Attr)) {
+    S.Diag(Attr.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::doublepump'";
+  } else if (Attr.getKind() == ParsedAttr::AT_IntelFPGASinglePump &&
```
Comment (premanandrao) [inline clang/lib/Sema/SemaDeclAttr.cpp:5227]: here too

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 15]
```diff
@@ -5192,6 +5220,16 @@ static void handleIntelFPGAPumpAttr(Sema &S, Decl *D, const ParsedAttr &Attr) {
...
+             checkDeprecatedSYCLAttributeSpelling(S, Attr)) {
+    S.Diag(Attr.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::singlepump'";
+  }
```
Comment (premanandrao) [inline clang/lib/Sema/SemaDeclAttr.cpp:5231]: and here

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 16]
```diff
@@ -5298,6 +5344,16 @@ static void handleOneConstantPowerTwoValueAttr(Sema &S, Decl *D,
   if (checkAttrMutualExclusion<IntelFPGARegisterAttr>(S, D, Attr))
     return;
 
+  if (Attr.getKind() == ParsedAttr::AT_IntelFPGABankWidth &&
+      checkDeprecatedSYCLAttributeSpelling(S, Attr)) {
+    S.Diag(Attr.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::bankwidth'";
+  } else if (Attr.getKind() == ParsedAttr::AT_IntelFPGANumBanks &&
```
Comment (premanandrao) [inline clang/lib/Sema/SemaDeclAttr.cpp:5351]: here

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 17]
```diff
@@ -5298,6 +5344,16 @@ static void handleOneConstantPowerTwoValueAttr(Sema &S, Decl *D,
...
+             checkDeprecatedSYCLAttributeSpelling(S, Attr)) {
+    S.Diag(Attr.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::numbanks'";
+  }
```
Comment (premanandrao) [inline clang/lib/Sema/SemaDeclAttr.cpp:5355]: and here

File: clang/lib/Sema/SemaStmtAttr.cpp [hunk 18]
```diff
@@ -95,6 +103,31 @@ static Attr *handleIntelFPGALoopAttr(Sema &S, const ParsedAttr &A) {
     }
   }
 
+  if (A.getParsedKind() == ParsedAttr::AT_SYCLIntelFPGAII &&
```
Comment (premanandrao) [inline clang/lib/Sema/SemaStmtAttr.cpp:106]: Does this need to be getParsedKind() unlike the rest of the ones below which use getKind()?

File: clang/lib/Sema/SemaStmtAttr.cpp [hunk 19]
```diff
@@ -111,6 +144,11 @@ Attr *handleIntelFPGALoopAttr<SYCLIntelFPGADisableLoopPipeliningAttr>(
     return nullptr;
   }
 
+  if (checkDeprecatedSYCLLoopAttributeSpelling(S, A)) {
+    S.Diag(A.getLoc(), diag::note_spelling_suggestion)
+        << "'intel::disable_loop_pipelining'";
+  }
```
Comment (premanandrao) [inline clang/lib/Sema/SemaStmtAttr.cpp:150]: unneeded braces

File: clang/lib/Sema/SemaStmtAttr.cpp [hunk 20]
```diff
@@ -272,6 +310,10 @@ static Attr *handleIntelFPGAIVDepAttr(Sema &S, const ParsedAttr &A) {
     return nullptr;
   }
 
+  if (checkDeprecatedSYCLLoopAttributeSpelling(S, A)) {
+    S.Diag(A.getLoc(), diag::note_spelling_suggestion) << "'intel::ivdep'";
+  }
```
Comment (premanandrao) [inline clang/lib/Sema/SemaStmtAttr.cpp:315]: here too

File: clang/test/SemaSYCL/intel-restrict.cpp [hunk 21]
```diff
@@ -24,7 +24,7 @@ int main() {
   // CHECK-LABEL: FunctionDecl {{.*}}test_kernel2
   // CHECK:       SYCLIntelKernelArgsRestrictAttr
   kernel<class test_kernel2>(
-      []() [[intel::kernel_args_restrict]] {});
+      []() [[intel::kernel_args_restrict]]{});
```
Comment (premanandrao) [inline clang/test/SemaSYCL/intel-restrict.cpp:27]: unrelated change

File: clang/lib/Sema/SemaStmtAttr.cpp (hunk 12 above)
Comment (smanna12) [reply clang/lib/Sema/SemaStmtAttr.cpp:129]: Thanks Mariya.

**Review by premanandrao — APPROVED** [review]

**Comment by erichkeane** [comment]
This looks fine as it is, though if we find ourselves wanting to do this MORE (that is, deprecating spellings), I think we should have a very significant look into adding a 'deprecated spelling' tag into Attr.td that gets taken care of automatically in the attribute infrastructure.  I don't know of any attributes (@AaronBallman ??) that currently have any spelling 'deprecated' in our upstream, but it is likely a good idea to keep an eye out for that opportunity.

**Comment by AaronBallman** [comment]
> This looks fine as it is, though if we find ourselves wanting to do this MORE (that is, deprecating spellings), I think we should have a very significant look into adding a 'deprecated spelling' tag into Attr.td that gets taken care of automatically in the attribute infrastructure. I don't know of any attributes (@AaronBallman ??) that currently have any spelling 'deprecated' in our upstream, but it is likely a good idea to keep an eye out for that opportunity.

We have a few attributes that have deprecated spellings. For instance, the `swift_wrapper` spelling in https://github.com/llvm/llvm-project/blob/master/clang/include/clang/Basic/Attr.td#L2179 is a deprecated spelling that was recently added. IIRC, some of the thread safety annotations also have alternate spellings that may be deprecated-ish (or may just be discouraged as older spellings, I'd have to research it).

**Review by Fznamznon — APPROVED** [review]

**Comment by smanna12** [comment]
Thank you everyone for the reviews.

**Comment by smanna12** [comment]
@bader, All BuildBot tests are completed and this PR is ready for your review/merge. Thanks.

**Comment by bader** [comment]
Thanks!
